core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java

```diff
@@ -387,4 +389,190 @@ public void dataFilesCleanup() throws IOException {
     Assert.assertTrue("FILE_A should be deleted", deletedFiles.contains(FILE_A.path().toString()));
     Assert.assertTrue("FILE_B should be deleted", deletedFiles.contains(FILE_B.path().toString()));
   }
+
+  /**
+   * Test on table below, and expiring the staged commit `B` using `expireOlderThan` API.
+   * Table: A - C
+   *          ` B (staged)
+   */
+  @Test
+  public void testWithExpiringDanglingStageCommit() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    TableMetadata base = readMetadata();
+    Snapshot snapshotA = base.snapshots().get(0);
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    Set<String> deletedFiles = new HashSet<>();
+
+    // Expire all commits including dangling staged snapshot.
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireOlderThan(snapshotB.timestampMillis() + 1)
+        .commit();
+
+    Set<String> expectedDeletes = new HashSet<>();
+    expectedDeletes.add(snapshotA.manifestListLocation());
+
+    // Files should be deleted of dangling staged snapshot
+    snapshotB.addedFiles().forEach(i -> {
+      expectedDeletes.add(i.path().toString());
+    });
+
+    // ManifestList should be deleted too
+    expectedDeletes.add(snapshotB.manifestListLocation());
+    snapshotB.manifests().forEach(file -> {
+      //Only the manifest of B should be deleted.
+      if (file.snapshotId() == snapshotB.snapshotId()) {
+        expectedDeletes.add(file.path());
+      }
+    });
+    Assert.assertSame("Files deleted count should be expected", expectedDeletes.size(), deletedFiles.size());
+    //Take the diff
+    expectedDeletes.removeAll(deletedFiles);
```

Comment:
Why does this remove files from the set instead of using `assertEqual`?

Reply:
The order was coming to be different, and I didn't wanted to rely on order.

Reply:
Comparing two sets would not rely on order.